components/dashboard/src/data/workspaces/listen-to-workspace-ws-messages.ts

```diff
@@ -51,12 +51,62 @@ export const useListenToWorkspacesWSMessages = () => {
     }, [organizationId, queryClient]);
 };
 
-export function watchWorkspaceStatus(
-    workspaceId: string | undefined,
-    cb: (response: WatchWorkspaceStatusResponse) => void,
-): Disposable {
+type WatchWorkspaceStatusCallback = (response: WatchWorkspaceStatusResponse) => Promise<void> | void;
+
+export function watchWorkspaceStatus(workspaceId: string | undefined, cb: WatchWorkspaceStatusCallback): Disposable {
     return stream<WatchWorkspaceStatusRequest>(
         (options) => workspaceClient.watchWorkspaceStatus({ workspaceId }, options),
         cb,
     );
 }
+
+const cachedCallbackInfoMap = new Map<string, { complete: WatchWorkspaceStatusCallback; priority: number }[]>();
+const cachedDisposables = new Map<string, Disposable>();
+
+export enum WatchWorkspaceStatusPriority {
+    StartWorkspacePage = 100,
+    SupervisorService = 50,
+}
+
+/**
+ * Registers multiple callbacks to receive the same workspace status update in order of priority.
+ *
+ * @param workspaceId The workspace ID to watch. If undefined, all workspaces are watched.
+ * @param priority The priority of the callback. Higher priority callbacks are executed and waited first.
+ * @param callback The callback to execute when a workspace status update is received. Only executed after previous callbacks.
+ */
+export function watchWorkspaceStatusInOrder(
```

Comment:
can we add a unit test for it to make sure that each client never miss events, and we proper release all disposables and so on? It does not need to use real API, can be mocked